services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/CompletedDirectoryDownload.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Represents a completed download directory transfer to Amazon S3. It can be used to track
+ * failed single file downloads.
+ *
+ * @see S3TransferManager#downloadDirectory(DownloadDirectoryRequest) 
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public final class CompletedDirectoryDownload implements CompletedDirectoryTransfer {
+
+    private final Collection<FailedFileDownload> failedTransfers;
+
+    private CompletedDirectoryDownload(DefaultBuilder builder) {
+        this.failedTransfers = Collections.unmodifiableCollection(
+            Validate.paramNotNull(builder.failedTransfers, "failedTransfers"));
+    }
+
+    @Override
+    public Collection<FailedFileDownload> failedTransfers() {
+        return failedTransfers;
+    }
```

Comment:
Discussion, non-blocking: IMO, it's more polite to return the most concrete abstract type, e.g. List instead of Collection here so that people can have constant-time random access to its members. I know that's controversial "because order doesn't matter" here, but I think it's more likely that a user would want to use failedTransfers().get(n) to get the n'th failure than a user would get confused about the order being "inconsistent" just because we returned a List instead of a Collection.

Reply:
Haha, @Bennett-Lynch and I had the same debate before! We can definitely open the discussion. The main reason I think Collection is better because it gives us flexibility to change the implementation (although it's very unlikely we will change it).
I think customers can still do `failedFileUploads.stream().collect(Collectors.toList())` if they want to iterate the list by index.

Reply:
If you've already convinced Bennett that this is the correct decision, then I don't want to make you have to debate this again.

Reply:
I was pro-`List` as well, and still somewhat am, but was mostly persuaded to use `Collection` since it allows us to use our [ConcurrentLinkedQueue](https://github.com/aws/aws-sdk-java-v2/blob/acda1370233a7f22fb19fb59faec86176f4d5a62/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelper.java#L105-L105) directly. I agree that ordering doesn't strictly matter here, but it still seems like a nice-to-have.

Reply:
@millems Everyone's opinion matters! Let's vote on it. :) 
@Bennett-Lynch We are still copying it to an ArrayList technically.